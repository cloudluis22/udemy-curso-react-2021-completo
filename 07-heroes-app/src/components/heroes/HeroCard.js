import React from 'react'
import { Link } from 'react-router-dom';

const heroImgs = require.context('../../assets/heroes', true);

export const HeroCard = ({
    id,
    superhero,
    alter_ego,
    publisher,
    first_appearance,
    characters
}) => {

    return (
        <div className="card mb-5 ms-5 pe-1 " style={ { maxWidth: 350 } }>
            <img className="card-img-top mt-3 rounded mx-auto d-block" src={ heroImgs(`./${id}.jpg`).default } alt="hero" style={ { maxWidth: 250, maxHeight: 300 } } />
            <div className="card-body">
                <h4 className="card-title"> {superhero} </h4>
                <h5 className="card-subtitle text-muted"> {alter_ego} </h5>

                {/* <p className="card-text mt-4"> <b> First Appearance: </b> {first_appearance} </p>
                {
                    (alter_ego !== characters) &&
                        <p className="card-text"> <b> Characters: </b> {characters} </p>
                } */}

                <Link to={`./hero/${id}`}>
                    ...MÃ¡s
                </Link>

            </div>
        </div>
    )
}
