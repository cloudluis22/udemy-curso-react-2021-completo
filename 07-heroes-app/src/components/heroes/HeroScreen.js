import React, { useMemo } from 'react'
import { Redirect, useParams } from 'react-router-dom'
import { getHeroById } from '../selectors/getHeroById';

const heroImgs = require.context('../../assets/heroes', true);

export const HeroScreen = ({history}) => {

    const { heroId } = useParams();

    const hero = useMemo( () => getHeroById(heroId), [heroId] );

    if (!hero) {
        return <Redirect to="/" />
    }

    const {
        id,
        superhero,
        publisher,
        alter_ego,
        first_appearance,
        characters
    } = hero;

    const handleReturn = () => {

        if (history.length <= 2) {
            if (publisher === 'DC Comics') {
                history.push('/dc');
            }
            else {
                history.push('/');
            }
        }
        else {
            history.goBack();
        }

    }

    return (
        <div className="row mt-5">

            <div className="col-4 animate__animated animate__fadeInLeftBig">
                <img className="img-thumbnail" src={ heroImgs(`./${id}.jpg`).default } alt={superhero} />
            </div>

            <div className="col-8">
                <h3> {superhero} </h3>
                <ul className="list-group list-group-flush">
                    <li className="list-group-item"> <b> Publisher: </b> {publisher} </li>
                    <li className="list-group-item"> <b> First Appearance: </b> {first_appearance} </li>
                    <li className="list-group-item"> <b> Alter Ego: </b> {alter_ego} </li>
                </ul>

                <h5 className="mt-3 ms-4"> Characters: </h5>
                <p className="ms-4"> {characters} </p>

                <button className="btn btn-outline-info ms-4" onClick={ handleReturn }> Return </button>
            </div>

        </div>
    )
}
