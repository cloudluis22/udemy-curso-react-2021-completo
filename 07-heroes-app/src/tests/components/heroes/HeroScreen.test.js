import React from 'react';
import { mount } from 'enzyme';
import { MemoryRouter, Redirect, Route } from 'react-router-dom';
import { HeroScreen } from '../../../components/heroes/HeroScreen';


describe('Pruebas en el componente funcional de "<HeroScreen/>".', () => {

    const historyMock = {
        length: 10,
        push: jest.fn(),
        goBack: jest.fn()
    }



    test('Debe de mostrar el componente redirect si no hay argumentos en el URL.', () => {

        const wrapper = mount(

            < MemoryRouter initialEntries={['/hero']} >
                <HeroScreen history={historyMock} />
            </MemoryRouter>

        );

        expect(wrapper).toMatchSnapshot();
        expect(wrapper.find(Redirect).exists()).toBe(true);

    });

    test('Debe de mostrar un héroe si el parámetro existe y se encuentra.', () => {

        const wrapper = mount(

            < MemoryRouter initialEntries={['/hero/marvel-spider']} >
                <Route path="/hero/:heroId" component={ HeroScreen } />
            </MemoryRouter>

        );

        expect(wrapper.find('.row').exists()).toBe(true);

    });

    test('Debe de regresar a la pantalla principal al darle click al botón con el push si no hay pags. anteriores.', () => {

        const history = {
            length: 1,
            push: jest.fn(),
            goBack: jest.fn()
        }

        const wrapper = mount(

            < MemoryRouter initialEntries={['/hero/marvel-spider']} >
                <Route path="/hero/:heroId" component={ () => <HeroScreen history={ history } />} />
            </MemoryRouter>

        );

        wrapper.find('button').prop('onClick')();

        expect(history.push).toHaveBeenCalledWith('/');
        expect(history.goBack).not.toHaveBeenCalled();

    });

    test('Debe de regresar a la pantalla anterior al darle click al botón de si sí hay pags. anteriores.', () => {

        const wrapper = mount(

            < MemoryRouter initialEntries={['/hero/marvel-spider']} >
                <Route path="/hero/:heroId" component={ () => <HeroScreen history={ historyMock } />} />
            </MemoryRouter>

        );

        wrapper.find('button').prop('onClick')();

        expect(historyMock.goBack).toHaveBeenCalled();
        expect(historyMock.push).not.toHaveBeenCalled();

    });

});