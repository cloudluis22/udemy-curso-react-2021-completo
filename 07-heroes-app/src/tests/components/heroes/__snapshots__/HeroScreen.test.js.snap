// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en el componente funcional de "<HeroScreen/>". Debe de mostrar el componente redirect si no hay argumentos en el URL. 1`] = `""`;
